src/Npgsql/Internal/Converters/FullTextSearch/TsQueryConverter.cs

```diff
@@ -0,0 +1,219 @@
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using NpgsqlTypes;
+using static NpgsqlTypes.NpgsqlTsQuery.NodeKind;
+
+// ReSharper disable once CheckNamespace
+namespace Npgsql.Internal.Converters;
+
+public class TsQueryConverter<T> : PgStreamingConverter<T>
+    where T : NpgsqlTsQuery
+{
+    // 1 (type) + 1 (weight) + 1 (is prefix search) + 2046 (max str len) + 1 (null terminator)
+    const int MaxSingleTokenBytes = 2050;
```

Comment:
Depending on what our official lower bound for our buffer sizes would be we could make this be a buffered converter. I personally would find a 4k minimum slightly too high, 2k I would be ok with but that won't fit.

Reply:
Note that this is the maximum for a single token, whereas a TsQuery contains multiple of these. So this supports buffered logic for chunks within the value, rather than for the entire value (which would be akin to making this a buffered converter.

But anyway, yeah, I had similar thoughts. I went the extra mile and removed all pre-buffering to make this fully streaming - no more minimum buffer requirements. Can you take a look to make sure I haven't missed anything?

Small note: it may be useful to distinguish the streaming APIs on reader/writer from the non-streaming ones (which throw); right now we have e.g. WriteByte next to WriteChars, and there's no indication that the former requires pre-buffering but the latter doesn't. Maybe just calling